### PR TITLE
Composer - replace threadgate modal with alf dialog

### PR DESCRIPTION
**Pull request by mozzius**
<table>
  <tr>
    <td>iOS default</td>
    <td>iOS with selections</td>
    <td>Web</td>
  </tr>
  <tr>
    <td>Simulator Screenshot - iPhone 15 - 2024-06-03 at 12 08 44</td>
    <td>Simulator Screenshot - iPhone 15 - 2024-06-03 at 12 08 50</td>
    <td>Screenshot 2024-06-03 at 12 15 16</td>
  </tr>
</table>

### CODE REVIEW

**Comment by gaearon** [comment]
Pressing Escape now tries to dismiss both dialogs (previously it was ignored). We need to fix this to only dismiss the inner one (like with GIF picker).

**Comment by gaearon** [comment]
also getting this weird spinner

Screenshot 2024-06-04 at 11 59 56

**Comment by mozzius** [comment]
That's it loading your lists, maybe we should change the layout so that it's less janky

**Comment by gaearon** [comment]
nah, it never loads. just stuck like this. there also no new network requests.

on main, there's no spinner.

**Review by gaearon — CHANGES_REQUESTED** [review]
- need to fix Escape handling
- distinguish no lists vs loading lists

File: src/view/com/composer/threadgate/ThreadgateBtn.tsx [hunk 1]
```diff
@@ -49,19 +51,168 @@ export function ThreadgateBtn({
...
+            label={_(msg`Mentioned users`)}
+            isSelected={!!threadgate.find(v => v.type === 'mention')}
+            onPress={() => onPressAudience({type: 'mention'})}
+          />
+          <Selectable
+            label={_(msg`Followed users`)}
+            isSelected={!!threadgate.find(v => v.type === 'following')}
+            onPress={() => onPressAudience({type: 'following'})}
+          />
+          {lists?.length ? (
```
[the remark this answers is not in the document]
Comment (mozzius) [reply src/view/com/composer/threadgate/ThreadgateBtn.tsx:152]: Oh whoops, missed that that was how it was doing that. Will fix

**Comment by gaearon** [comment]
The height is jumping when changing selection, probably due to borders appearing and disappearing:

https://github.com/bluesky-social/social-app/assets/810438/b842b4fc-da16-42fd-be9d-c23d543ff617

It would be good to stabilize those.

Also, I'm not a big fan of using the same color for selection and the primary action:

Screenshot 2024-06-04 at 15 00 05

Let's do a pass over these before getting it in.

**Comment by pfrazee** [comment]
Updated the buttons

|one|two|
|-|-|
|CleanShot 2024-06-24 at 11 34 13@2x|CleanShot 2024-06-24 at 11 34 39@2x|

File: src/components/WhoCanReply.tsx [hunk 2]
```diff
@@ -63,6 +71,53 @@ export function WhoCanReply({post, isThreadAuthor, style}: WhoCanReplyProps) {
...
+        const thread = res.data.thread
+        if (AppBskyFeedDefs.isThreadViewPost(thread)) {
+          const fetchedSettings = threadgateViewToSettings(
+            thread.post.threadgate,
+          )
+          return JSON.stringify(fetchedSettings) === JSON.stringify(newSettings)
+        }
+        return false
+      })
+      Toast.show('Thread settings updated')
```
Comment (surfdude29) [inline src/components/WhoCanReply.tsx:109]: Sorry to jump in, could the toast here and on L115 be marked for localization please? 🙏

**Comment by pfrazee** [comment]
@gaearon Interesting, is that 2-states solution preferable to using an effect?

**Comment by gaearon** [comment]
yeah, it avoids a "roundtrip", see here https://react.dev/reference/react/useState#storing-information-from-previous-renders :

>This pattern can be hard to understand and is usually best avoided. However, it’s better than updating state in an effect. When you call the set function during render, React will re-render that component immediately after your component exits with a return statement, and before rendering the children. This way, children don’t need to render twice. The rest of your component function will still execute (and the result will be thrown away). If your condition is below all the Hook calls, you may add an early return; to restart rendering earlier.

however i think i'll actually move the state down so that it gets torn down on dialog close

**Review by pfrazee — APPROVED** [review]
Excellent work @mozzius @gaearon

**Review by gaearon — APPROVED** [review]

**Pull request merged**